Sources/AsyncAlgorithms/Debounce/DebounceStorage.swift

```diff
@@ -0,0 +1,311 @@
+                                    precondition(error is CancellationError, "Received unexpected error \(error) in the Clock loop")
+                                    break loop
+                                }
+                            }
+                        }
+
+                        do {
+                            try await group.waitForAll()
+                        } catch {
+                            // The upstream sequence threw an error
+                            let action = self.lock.withLock {
+                                self.stateMachine.upstreamThrew(error)
+                            }
+
+                            switch action {
+                            case .resumeContinuationWithErrorAndCancelTaskAndUpstreamContinuation(
+                                let downstreamContinuation,
+                                let error,
+                                let task,
+                                let upstreamContinuation,
+                                let clockContinuation
+                            ):
+                                upstreamContinuation?.resume(throwing: CancellationError())
+                                clockContinuation?.resume(throwing: CancellationError())
+
+                                task.cancel()
+
+                                downstreamContinuation.resume(throwing: error)
+
+                            case .cancelTaskAndClockContinuation(
+                                let task,
+                                let clockContinuation
+                            ):
+                                clockContinuation?.resume(throwing: CancellationError())
+                                task.cancel()
+
+                            case .none:
+                                break
+                            }
+
+                            group.cancelAll()
+                        }
+                    }
+                }
+
+                self.stateMachine.taskStarted(task)
+            }
+        }
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.lock.withLock { self.stateMachine.iteratorDeinitialized() }
+
+        switch action {
+        case .cancelTaskAndUpstreamAndClockContinuations(
+            let task,
+            let upstreamContinuation,
+            let clockContinuation
+        ):
+            upstreamContinuation?.resume(throwing: CancellationError())
+            clockContinuation?.resume(throwing: CancellationError())
+
+            task.cancel()
+
+        case .none:
+            break
+        }
+    }
+
+    func next() async rethrows -> Element? {
+        // We need to handle cancellation here because we are creating a continuation
+        // and because we need to cancel the `Task` we created to consume the upstream
+        return try await withTaskCancellationHandler {
+            // We always suspend since we can never return an element right away
+
+            self.lock.lock()
```

Comment:
this is the only straddling lock, I wonder if we can wrap this up into something within the managed critical state. so that for future use it is ONLY the safe edge case being used here.